Bottom: 62f5955f249a162e25f0bb0d64c68216f943fade
Top:    69be1c7205658c882e5af23d2e838078d4c45c1d
Author: Jason Wang <jasowang@redhat.com>
Date:   2017-05-28 14:41:18 +0800

Refresh of new-road

---

diff --git a/drivers/vhost/vhost.c b/drivers/vhost/vhost.c
index 281b5fd..eb0c97c 100644
--- a/drivers/vhost/vhost.c
+++ b/drivers/vhost/vhost.c
@@ -2422,14 +2422,14 @@ int vhost_prefetch_heads(struct vhost_virtqueue *vq,
 
 	for (i = 0; i < total; i++) {
 		if (unlikely(vhost_get_avail(vq, heads[i].id,
-		    &vq->avail->ring[last_avail_idx & (vq->num - 1)]))) {
+							&vq->avail->ring[(last_avail_idx
+		+ i)& (vq->num - 1)]))) {
 			vq_err(vq, "Failed to get descriptors\n");
 			return -EFAULT;
 		}
 		heads[i].len = peek(vq, i);
-		++last_avail_idx;
+		vhost_add_used_elem(vq, heads[i].id, heads[i].len, i);
 	}
-	__vhost_add_used_n(vq, heads, total);
 
 	/* Only get avail ring entries after they have been exposed by guest. */
 	smp_rmb();
